README.md
# fs-git [![Circle CI](https://circleci.com/gh/vvakame/fs-git.png?style=badge)](https://circleci.com/gh/vvakame/fs-git) [![Build status](https://ci.appveyor.com/api/projects/status/v48xurjhxh0evabl/branch/master?svg=true)](https://ci.appveyor.com/project/vvakame/fs-git/branch/master)

file system like api for git repository

see [test/mainSpec.ts](https://github.com/vvakame/fs-git/blob/master/test/mainSpec.ts#L37)


CHANGELOG.md
<a name="1.0.2"></a>
## [1.0.2](https://github.com/vvakame/fs-git/compare/v1.0.1...v1.0.2) (2017-06-01)


### Bug Fixes

* **fs-git:** fix command injection ([eb5f70e](https://github.com/vvakame/fs-git/commit/eb5f70e))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/vvakame/fs-git/compare/v1.0.0...v1.0.1) (2016-02-21)

v1.0.0 is deprecated. it is not contains lib/*.js . sorry!

### Features

* **ci:** update node.js version 4.3 in CI ([6182887](https://github.com/vvakame/fs-git/commit/6182887))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/vvakame/fs-git/compare/0.5.1...v1.0.0) (2016-02-21)


### Features

* **desp:** update dependencies ([783c56e](https://github.com/vvakame/fs-git/commit/783c56e))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/vvakame/fs-git/compare/0.5.0...v0.5.1) (2015-09-22)


### Bug Fixes

* **fs-git:** rename lib/fs.ts to lib/index.ts ([c1ed01c](https://github.com/vvakame/fs-git/commit/c1ed01c))



<a name="0.4.1"></a>
## [0.4.1](https://github.com/vvakame/fs-git/compare/0.4.1...v0.4.1) (2015-09-22)


### Bug Fixes

* **fs-git:** fix test before expression ([d9a200f](https://github.com/vvakame/fs-git/commit/d9a200f))
* **fs-git:** set mocha timeout to 10000ms ([859b307](https://github.com/vvakame/fs-git/commit/859b307))

### Features

* **ci:** add appveyor.yml ([53d949e](https://github.com/vvakame/fs-git/commit/53d949e))
* **fs-git:** update dependencies & switch to tsconfig.json oriented style ([bf4e25f](https://github.com/vvakame/fs-git/commit/bf4e25f))



<a name="0.4.1"></a>
### 0.4.1 (2015-02-08)


#### Bug Fixes

* **fs-git:** change maxBuffer size 200KB to 1MB ([a3ceaede](https://github.com/vvakame/fs-git/commit/a3ceaede79018365e36895d1a5d7f61072210e4d))


#### Features

* **fs-git:** move feature of es6 polyfill main code to test code ([efa60277](https://github.com/vvakame/fs-git/commit/efa602771346b147db9c4e9703da47c4c8327b58))


<a name="0.4.0"></a>
## 0.4.0 (2015-02-08)


#### Features

* **fs-git:** add showRef method ([f94480a8](https://github.com/vvakame/fs-git/commit/f94480a846bc660161aa181bcdae971c47082482))


<a name="0.3.0"></a>
## 0.3.0 (2015-02-07)


#### Features

* **deps:** add grunt-conventional-changelog ([d435a9ee](https://github.com/vvakame/fs-git/commit/d435a9ee34e09764791e761079f7640f9f0d8c6c))
* **fs-git:** add file method to FSGit ([a2aa3e97](https://github.com/vvakame/fs-git/commit/a2aa3e9782e046e30fca4631b79b01ad60652212))


#### Breaking Changes

* rename FSGit.filelist to FSGit.fileList
 ([268ba3f0](https://github.com/vvakame/fs-git/commit/268ba3f088a5351e5267e8fa23f30f5fff447722))


test/mainSpec.ts
require("es6-promise").polyfill();

try {
    // optional
    require("source-map-support").install();
} catch (e) {
}

import * as assert from "power-assert";
import * as fsgit from "../lib/";

import * as fs from "fs";
import * as child_process from "child_process";

describe("fs-git", () => {
    before(done => {
        if (fs.existsSync("./test/fixture/vanilla")) {
            done();
            return;
        }
        child_process.exec("git clone test/fixture/bare.git test/fixture/vanilla", err => {
            if (err) {
                done(err);
            }
            child_process.exec("git checkout -b develop origin/develop", { cwd: "test/fixture/vanilla" }, err => {
                done(err);
            });
        });
    });

    let matrix: { name: string; gitDir: string; }[] = [
        {
            name: "bare repository",
            gitDir: "./test/fixture/bare.git",
        },
        {
            name: "cloned repository",
            gitDir: "./test/fixture/vanilla/.git",
        },
    ];
    matrix.forEach(dirInfo => {
        it("show refs in " + dirInfo.name, () => {
            return fsgit.open(dirInfo.gitDir, "master").then(fs => {
                fs.showRef().then(refs => {
                    assert(refs.length !== 0);
                    refs.forEach(ref => {
                        assert(/^[0-9a-f]+$/.test(ref.ref));
                        assert(/^refs\//.test(ref.name));
                    });
                });
            });
        });

        it("open master branch in " + dirInfo.name, () => {
            return fsgit.open(dirInfo.gitDir, "master").then(fs => {
                let fileList = fs.fileList().then(files => {
                    assert(files.length !== 0);
                    assert(files.filter(file => file.path === "master-branch-only.js").length === 1);
                });
                let file = fs.file("subdir2/master-branch-only.js").then(fileInfo => {
                    assert(fileInfo.type === "blob");
                });
                let readFile = fs.readFile("subdir2/master-branch-only.js", { encoding: "utf8" }).then(content => {
                    assert(typeof content === "string");
                });
                let exists1 = fs.exists("subdir2/master-branch-only.js").then(exists => {
                    assert(exists === true);
                });
                let exists2 = fs.exists("develop-branch-only.js").then(exists => {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });

        it("open develop branch in " + dirInfo.name, () => {
            return fsgit.open(dirInfo.gitDir, "develop").then(fs => {
                let fileList = fs.fileList().then(files => {
                    assert(files.length !== 0);
                    assert(files.filter(file => file.path === "develop-branch-only.js").length === 1);
                });
                let file = fs.file("subdir/develop-branch-only.js").then(fileInfo => {
                    assert(fileInfo.type === "blob");
                });
                let readFile = fs.readFile("subdir/develop-branch-only.js", { encoding: "utf8" }).then(content => {
                    assert(typeof content === "string");
                });
                let exists1 = fs.exists("develop-branch-only.js").then(exists => {
                    assert(exists === true);
                });
                let exists2 = fs.exists("master-branch-only.js").then(exists => {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });

        it("open test-tag tag in " + dirInfo.name, () => {
            return fsgit.open(dirInfo.gitDir, "test-tag").then(fs => {
                let fileList = fs.fileList().then(files => {
                    assert(files.length !== 0);
                    assert(files.filter(file => file.path === "subdir/test.txt").length === 1);
                });
                let file = fs.file("subdir/test.txt").then(fileInfo => {
                    assert(fileInfo.type === "blob");
                });
                let readFile = fs.readFile("subdir/test.txt", { encoding: "utf8" }).then(content => {
                    assert(typeof content === "string");
                });
                let exists1 = fs.exists("subdir/test.txt").then(exists => {
                    assert(exists === true);
                });
                let exists2 = fs.exists("master-branch-only.js").then(exists => {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });

        it("open specific ref in " + dirInfo.name, () => {
            return fsgit.open(dirInfo.gitDir, "b41735").then(fs => {
                let fileList = fs.fileList().then(files => {
                    assert(files.length !== 0);
                    assert(files.filter(file => file.path === "subdir/test.txt").length === 1);
                });
                let file = fs.file("subdir/test.txt").then(fileInfo => {
                    assert(fileInfo.type === "blob");
                });
                let readFile = fs.readFile("subdir/test.txt", { encoding: "utf8" }).then(content => {
                    assert(typeof content === "string");
                });
                let exists1 = fs.exists("subdir/test.txt").then(exists => {
                    assert(exists === true);
                });
                let exists2 = fs.exists("master-branch-only.js").then(exists => {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });
    });
});


test/main_spec.ts
/// <reference path="../typings/node/node.d.ts" />
/// <reference path="../typings/es6-promise/es6-promise.d.ts" />

/// <reference path="../typings/mocha/mocha.d.ts" />
/// <reference path="../typings/power-assert/power-assert.d.ts" />

require("es6-promise").polyfill();

try {
    // optional
    require("source-map-support").install();
} catch (e) {
}

import assert = require("power-assert");
import fsgit = require("../lib/fs");

import fs = require("fs");
import child_process = require("child_process");

describe("fs-git", ()=> {
    before(done => {
        if (fs.existsSync("./test/fixture/vanilla")) {
            done();
            return;
        }
        child_process.exec("git clone test/fixture/bare.git test/fixture/vanilla", err=> {
            if (err) {
                done(err);
            }
            child_process.exec("git checkout -b develop origin/develop", {cwd: "test/fixture/vanilla"}, err=> {
                done(err);
            });
        });
    });

    var matrix:{name: string; gitDir: string; }[] = [
        {
            name: "bare repository",
            gitDir: "./test/fixture/bare.git"
        },
        {
            name: "cloned repository",
            gitDir: "./test/fixture/vanilla/.git"
        }
    ];
    matrix.forEach(dirInfo => {
        it("show refs in " + dirInfo.name, ()=> {
            return fsgit.open(dirInfo.gitDir, "master").then(fs => {
                fs.showRef().then(refs => {
                    assert(refs.length !== 0);
                    refs.forEach(ref => {
                        assert(/^[0-9a-f]+$/.test(ref.ref));
                        assert(/^refs\//.test(ref.name));
                    });
                });
            });
        });

        it("open master branch in " + dirInfo.name, ()=> {
            return fsgit.open(dirInfo.gitDir, "master").then(fs => {
                var fileList = fs.fileList().then(files=> {
                    assert(files.length !== 0);
                    assert(files.filter(file=> file.path === "master-branch-only.js").length === 1);
                });
                var file = fs.file("subdir2/master-branch-only.js").then(fileInfo=> {
                    assert(fileInfo.type === "blob");
                });
                var readFile = fs.readFile("subdir2/master-branch-only.js", {encoding: "utf8"}).then(content=> {
                    assert(typeof content === "string");
                });
                var exists1 = fs.exists("subdir2/master-branch-only.js").then(exists=> {
                    assert(exists === true);
                });
                var exists2 = fs.exists("develop-branch-only.js").then(exists=> {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });

        it("open develop branch in " + dirInfo.name, ()=> {
            return fsgit.open(dirInfo.gitDir, "develop").then(fs => {
                var fileList = fs.fileList().then(files=> {
                    assert(files.length !== 0);
                    assert(files.filter(file=> file.path === "develop-branch-only.js").length === 1);
                });
                var file = fs.file("subdir/develop-branch-only.js").then(fileInfo=> {
                    assert(fileInfo.type === "blob");
                });
                var readFile = fs.readFile("subdir/develop-branch-only.js", {encoding: "utf8"}).then(content=> {
                    assert(typeof content === "string");
                });
                var exists1 = fs.exists("develop-branch-only.js").then(exists=> {
                    assert(exists === true);
                });
                var exists2 = fs.exists("master-branch-only.js").then(exists=> {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });

        it("open test-tag tag in " + dirInfo.name, ()=> {
            return fsgit.open(dirInfo.gitDir, "test-tag").then(fs => {
                var fileList = fs.fileList().then(files=> {
                    assert(files.length !== 0);
                    assert(files.filter(file=> file.path === "subdir/test.txt").length === 1);
                });
                var file = fs.file("subdir/test.txt").then(fileInfo=> {
                    assert(fileInfo.type === "blob");
                });
                var readFile = fs.readFile("subdir/test.txt", {encoding: "utf8"}).then(content=> {
                    assert(typeof content === "string");
                });
                var exists1 = fs.exists("subdir/test.txt").then(exists=> {
                    assert(exists === true);
                });
                var exists2 = fs.exists("master-branch-only.js").then(exists=> {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });

        it("open specific ref in " + dirInfo.name, ()=> {
            return fsgit.open(dirInfo.gitDir, "b41735").then(fs => {
                var fileList = fs.fileList().then(files=> {
                    assert(files.length !== 0);
                    assert(files.filter(file=> file.path === "subdir/test.txt").length === 1);
                });
                var file = fs.file("subdir/test.txt").then(fileInfo=> {
                    assert(fileInfo.type === "blob");
                });
                var readFile = fs.readFile("subdir/test.txt", {encoding: "utf8"}).then(content=> {
                    assert(typeof content === "string");
                });
                var exists1 = fs.exists("subdir/test.txt").then(exists=> {
                    assert(exists === true);
                });
                var exists2 = fs.exists("master-branch-only.js").then(exists=> {
                    assert(exists === false);
                });
                return Promise.all([fileList, file, readFile, exists1, exists2]);
            });
        });
    });
});


fs-git.d.ts
// Generated by dts-bundle v0.2.0

declare module 'fs-git' {
    export function open(path: string, ref?: string): Promise<FSGit>;
    export class FSGit {
        path: string;
        ref: string;
        constructor(path: string, ref?: string);
        file(path: string): Promise<FileInfo>;
        fileList(): Promise<FileInfo[]>;
        showRef(): Promise<RefInfo[]>;
        readFile(path: string): Promise<Buffer>;
        readFile(path: string, opts: {
            encoding: string;
        }): Promise<string>;
        exists(path: string): Promise<boolean>;
        revParse(ref: string): Promise<string>;
        _lsTree(ref?: string, path?: string): Promise<FileInfo[]>;
        _buildCommand(...args: string[]): string;
    }
    export interface FileInfo {
        gitDir: string;
        ref: string;
        permission: string;
        type: string;
        hash: string;
        path: string;
    }
    export interface RefInfo {
        gitDir: string;
        ref: string;
        name: string;
    }
}



Gruntfile.js
module.exports = function (grunt) {
    grunt.initConfig({
        pkg: grunt.file.readJSON('package.json'),
        conventionalChangelog: {
          options: {
            changelogOpts: {
              // conventional-changelog options go here
              preset: "angular"
           },
           context: {
              // context goes here
           },
           gitRawCommitsOpts: {
              // git-raw-commits options go here
           },
           parserOpts: {
              // conventional-commits-parser options go here
           },
           writerOpts: {
              // conventional-changelog-writer options go here
           }
          },
          release: {
            src: "CHANGELOG.md"
          }
        }
    });

    require('load-grunt-tasks')(grunt);
};


lib/index.ts
import * as child_process from "child_process";

export function open(path: string, ref?: string): Promise<FSGit> {
    return Promise.resolve(new FSGit(path, ref));
}

let maxBuffer = 1 * 1024 * 1024; // Node.js default 200*1024

export class FSGit {
    constructor(public path: string, public ref = "master") {
    }

    file(path: string): Promise<FileInfo> {
        return this._lsTree(this.ref, path)
            .then(fileList => {
                let fileInfo = fileList.filter(fileInfo => fileInfo.path === path)[0];
                if (fileInfo) {
                    return fileInfo;
                } else {
                    throw new Error(`${path} is not exists`);
                }
            });
    }

    fileList(): Promise<FileInfo[]> {
        return this._lsTree(this.ref, ".");
    }

    showRef(): Promise<RefInfo[]> {
        let command = this._buildCommand("show-ref");
        return new Promise((resolve: (value: RefInfo[]) => void, reject: (error: any) => void) => {
            child_process.execFile(command.base, command.args, { encoding: "buffer", maxBuffer: maxBuffer }, (error, stdout, stderr) => {
                if (error) {
                    reject(error);
                } else {
                    let list = stdout.toString("utf8").split("\n").filter(line => !!line);
                    let resultList: RefInfo[] = list.map(str => {
                        let columns = str.split(" ", 2);
                        return {
                            gitDir: this.path,
                            ref: columns[0],
                            name: columns[1],
                        };
                    });
                    resolve(resultList);
                }
            });
        });
    }

    readFile(path: string): Promise<Buffer>;

    readFile(path: string, opts: { encoding: string; }): Promise<string>;

    readFile(path: string, opts?: { encoding: string; }): Promise<any> {
        let command = this._buildCommand("show", this.ref + ":" + path);
        return new Promise((resolve: (value: any) => void, reject: (error: any) => void) => {
            child_process.execFile(command.base, command.args, { encoding: "buffer", maxBuffer: maxBuffer }, (error, stdout, stderr) => {
                if (error) {
                    reject(error);
                } else {
                    if (opts && opts.encoding) {
                        resolve(stdout.toString(opts.encoding));
                    } else {
                        resolve(stdout);
                    }
                }
            });
        });
    }

    exists(path: string): Promise<boolean> {
        return this.fileList().then(list => list.some(data => data.path === path));
    }

    revParse(ref: string): Promise<string> {
        let command = this._buildCommand("rev-parse", ref);

        return new Promise((resolve: (value?: any) => void, reject: (error: any) => void) => {
            child_process.execFile(command.base, command.args, { encoding: "buffer", maxBuffer: maxBuffer }, (error, stdout, stderr) => {
                if (error) {
                    console.log(command);
                    reject(error);
                } else {
                    let list = stdout.toString("utf8").split("\n").filter(str => str.length !== 0);
                    resolve(list[0]);
                }
            });
        });
    }

    _lsTree(ref = this.ref, path = "."): Promise<FileInfo[]> {
        return this.revParse(ref).then(ref => {
            let command = this._buildCommand("ls-tree", "-r", "-z", "--full-name", ref, path);
            return new Promise((resolve: (value: FileInfo[]) => void, reject: (error: any) => void) => {
                child_process.execFile(command.base, command.args, { encoding: "buffer", maxBuffer: maxBuffer }, (error, stdout, stderr) => {
                    if (error) {
                        reject(error);
                    } else {
                        let list = stdout.toString("utf8").split("\0").filter(str => str.length !== 0);
                        let resultList: FileInfo[] = list.map(str => {
                            let matches = str.match(/^([0-9]+)\s([^\s]+)\s([0-9a-f]+)\t(.+)$/);
                            return {
                                gitDir: this.path,
                                ref: ref,
                                permission: matches[1],
                                type: matches[2],
                                hash: matches[3],
                                path: matches[4],
                            };
                        });
                        resolve(resultList);
                    }
                });
            });
        });
    }

    _buildCommand(...args: string[]): { base: string; args: string[]; } {
        return {
            base: "git",
            args: [`--git-dir=${this.path}`, ...args],
        };
    }
}

export interface FileInfo {
    gitDir: string;
    ref: string;
    permission: string;
    type: string;
    hash: string;
    path: string;
}

export interface RefInfo {
    gitDir: string;
    ref: string;
    name: string;
}


lib/fs.ts
"use strict";

// if you use Node.js 0.10, you need exec `require("es6-promise").polyfill();`

import * as child_process from "child_process";

export function open(path:string, ref?:string):Promise<FSGit> {
    "use strict";

    return Promise.resolve(new FSGit(path, ref));
}

var maxBuffer = 1 * 1024 * 1024; // Node.js default 200*1024

export class FSGit {
    constructor(public path:string, public ref = "master") {
    }

    file(path:string):Promise<FileInfo> {
        return this._lsTree(this.ref, path)
            .then(fileList => {
                var fileInfo = fileList.filter(fileInfo => fileInfo.path === path)[0];
                if (fileInfo) {
                    return fileInfo;
                } else {
                    throw new Error(`${path} is not exists`);
                }
            });
    }

    fileList():Promise<FileInfo[]> {
        return this._lsTree(this.ref, ".");
    }

    showRef():Promise<RefInfo[]> {
        var command = this._buildCommand("show-ref");
        return new Promise((resolve:(value:RefInfo[])=>void, reject:(error:any)=>void) => {
            child_process.exec(command, {maxBuffer: maxBuffer}, (error, stdout, stderr)=> {
                if (error) {
                    reject(error);
                } else {
                    var list = stdout.toString("utf8").split("\n").filter(line => !!line);
                    var resultList:RefInfo[] = list.map(str=> {
                        var columns = str.split(" ", 2);
                        return {
                            gitDir: this.path,
                            ref: columns[0],
                            name: columns[1]
                        };
                    });
                    resolve(resultList);
                }
            });
        });
    }

    readFile(path:string):Promise<Buffer>;

    readFile(path:string, opts:{encoding: string;}):Promise<string>;

    readFile(path:string, opts?:{encoding: string;}):Promise<any> {
        var command = this._buildCommand("show", this.ref + ":" + path);
        return new Promise((resolve:(value:any)=>void, reject:(error:any)=>void) => {
            child_process.exec(command, {maxBuffer: maxBuffer}, (error, stdout, stderr)=> {
                if (error) {
                    reject(error);
                } else {
                    if (opts && opts.encoding) {
                        resolve(stdout.toString(opts.encoding));
                    } else {
                        resolve(stdout);
                    }
                }
            });
        });
    }

    exists(path:string):Promise<boolean> {
        return this.fileList().then(list=> list.some(data => data.path === path));
    }

    revParse(ref:string):Promise<string> {
        var command = this._buildCommand("rev-parse", ref);

        return new Promise((resolve:(value?:any)=>void, reject:(error:any)=>void)=> {
            child_process.exec(command, {maxBuffer: maxBuffer}, (error, stdout, stderr)=> {
                if (error) {
                    console.log(command);
                    reject(error);
                } else {
                    var list = stdout.toString("utf8").split("\n").filter(str => str.length !== 0);
                    resolve(list[0]);
                }
            });
        });
    }

    _lsTree(ref = this.ref, path = "."):Promise<FileInfo[]> {
        return this.revParse(ref).then(ref=> {
            var command = this._buildCommand("ls-tree", "-r", "-z", "--full-name", ref, path);
            return new Promise((resolve:(value:FileInfo[])=>void, reject:(error:any)=>void) => {
                child_process.exec(command, {maxBuffer: maxBuffer}, (error, stdout, stderr)=> {
                    if (error) {
                        reject(error);
                    } else {
                        var list = stdout.toString("utf8").split("\0").filter(str => str.length !== 0);
                        var resultList:FileInfo[] = list.map(str=> {
                            var matches = str.match(/^([0-9]+)\s([^\s]+)\s([0-9a-f]+)\t(.+)$/);
                            return {
                                gitDir: this.path,
                                ref: ref,
                                permission: matches[1],
                                type: matches[2],
                                hash: matches[3],
                                path: matches[4]
                            };
                        });
                        resolve(resultList);
                    }
                });
            });
        });
    }

    _buildCommand(...args:string[]):string {
        return `git --git-dir=${this.path} ${args.join(" ")}`;
    }
}

export interface FileInfo {
    gitDir: string;
    ref: string;
    permission: string;
    type: string;
    hash: string;
    path: string;
}

export interface RefInfo {
    gitDir: string;
    ref: string;
    name: string;
}
